### ISSUE TITLE
Another Possible Approach to Projected Prices and Volatility in Discovery Period

### ISSUE THREAD

**Issue by KelleyDrake** [ddrake/Integrated_Farm_Budget_Tool#223]
What we are doing now for prices and volatilities prior to and during discovery is fine.  But I have a suggestion that would make this better.  Not anything immediate required.    Rather a suggestion for a longer-term improvement.

Currently, prior to discovery period we use prior year's volatility and current harvest price as our inputs.  That is fine.   Then during the discovery period, we use an average of the closing prices within the discovery period to date and prior year's volatility.   At the end of the discovery period, we revert to the final actual numbers produced by the RMA for projected price and volatility factor.    As we know, the RMA has a discover price tool https://prodwebnlb.rma.usda.gov/apps/PriceDiscovery/GetPrices/YourPrice

Each day within the price discovery period, the RMA updates this tool and publishes the average to date projected price (which should match our calculation) and the average to date of last 5 days volatilities to produce an estimated volatility factor.   Currently, for example, after only two days into discovery, soybeans in Illinois on the RMA tool shows $11.78 average price to date and an average volatility factor of .15.  Our IFBT model shows the same $11.78 average price to date but uses prior year volatility which was .13. 

That difference in volatility factor between the RMA's at .15 and IFBT's at .13 doesn't sound like much but I believe that that difference has a pretty significant impact on premiums (perhaps even proportionate -- which would mean that we woud be currently understating premiums by 15%).    This, of course, will all get reconciled at the end of the discovery period.   But for farmer planning purposes during the discovery period it would be better to be using the .15 volatility factor (or whatever the latest RMA numbers was as of the latest date) instead of prior year's .13 volatility factor.    For 2024 so far, the volatility factors are pretty close (.15 vs .13).  But in some years current volatility factors as calculated by the RMA are vastly different from the prior year's volatility.

So here is my question...  Instead of calculating projected price within the discovery period and using last year's volatility factor, would it be possible to simply "grab" the actual published data from the RMA's for these two items?   We would no longer have to calculate average prices, or attempt to create a formula to create the estimated volatility factor and average the last 5 days of those.   Since we would be "grabbing" their actual data, we wouldn't have to worry about our data lining up exactly with theirs.   

 Instead, perhaps we could just go out an grab the running estimates right at the source -- data that the RMA themselves are publishing.   I'm not sure how to access this daily RMA data (other than going out and running the Your Price application which obviously is not scalable).   But somewhere inside the RMA's databases, they are computing and then publishing all of these various discovery data for average prices and average volatility during the discovery period.   Is that database something that we could access directly and become our default? 

I know this change is likely not trivial from a programming perspective and testing perspective.  And therefore is something to consider for the future -- rather than any near-term requirement.   This change would also have potential implications for "time travel" which might further complicate things.  What would happen if the user time-traveled back to a date during a discovery period?   But since we are starting the very important central grain belt discovery periods for corn and soybeans, I thought that it might be something we'd like to consider.   Hope that makes sense?

**Comment by ddrake** [comment]
I like this idea in principle.  In practice, this means developing a script to fill the RMA price discovery form for each supported commodity, type, practice, state and sales closing date (e.g. for TX), then scrape the projected price and volatility from the page displayed after submitting the form.  This would be similar to the script I developed to scrape premiums, but simpler.  I believe the script could update directly the ext_price table, which contains RMA prices, volatilities and yields.  It could be scheduled to run daily from Dec 15 to Mar 1.  As you mentioned, this would simplify the application and database code by removing the price averaging.

The main obstacles are to handle cases when the RMA site times out or gives an error response and handle the scrolling of the state field.  I think these should not be too difficult.

**Comment by ddrake** [comment]
We might also want to unschedule (suspend) the regular RMA update during that period (Dec 15 - Mar 1) to prevent the new data stream from being overwritten.  This may not be necessary if the regular RMA update script does not replace numeric values with NULL values, as I believe is the case.

**Comment by ddrake** [comment]
Time travel back to the middle of the discovery period would not match original values, but traveling back to before the discovery start would (check this).

**Comment by ddrake** [comment]
Comment by Kelley:  This is not a big deal but might also be solved by the possible enhancement of going directly to the RMA for both volatility and running discovered prices during the discovery period. As of 2/20 we had an average discovered corn price of $4.69 as compared to RMA of $4.70. We had an average discovered bean price of $11.64 as compared to RMA of $11.66. So our average prices are not exactly matching RMAs. Not a big deal but wanted you to be aware of this minor price discrepancy. If we were, at some point in the future, to make the previously discussed enhancement relating to pulling down RMA's data directly (instead of calculating the prices and using last year's volatility), we would match them exactly. FYI IFBT is using .18 volatility (last year's) for corn and RMA's latest calculation is .19. IFBT's soybean volatility is .13 (last year's) as compared to .15 for RMA. Nothing urgent to do here. But if we could ultimately match RMA exactly during discovery periods by "grabbling" their data, we'd not have these minor discrepancies on average prices to date or volatility factors that are resident in IFBT currently.

**Comment by ddrake** [comment]
Yes, I agree that we should go forward with the alternative approach, and not waste time trying to resolve the price issue, which will become moot.
